src/bash/patches/0001-Add-plugin-support-to-bash.patch

```diff
@@ -0,0 +1,804 @@
++    new_plugin_node->plugin_uninit = plugin_uninit;
++
++#ifdef DEBUG
++    itrace("Plugin: append plugin node %p to global list %p\n", new_plugin_node, global_plugin_list);
++#endif
++    
++    /* Find the pointer to the latest plugin node's 'next' field */
++    PLUGIN_NODE **current_plugin_node = &global_plugin_list;
++    while (*current_plugin_node != NULL) {
++        current_plugin_node = &((*current_plugin_node)->next);
++
++#ifdef DEBUG
++        itrace("Plugin: founded next plugin node: %p\n", *current_plugin_node);
++#endif
++    }
++    
++    /* append new plugin to tail node */
++    *current_plugin_node = new_plugin_node;
++
++#ifdef DEBUG
++    itrace("Plugin: append new plugin node %p to %p\n", new_plugin_node, current_plugin_node);
++#endif
++
++    return PLUGIN_LOAD_SUCCESS;
++}
++
++
++/* Load plugin by plugin path */
++void
++try_load_plugin_by_path(plugin_path)
++     const char *plugin_path;
++{
++    /* Plugin handle */
++    void *plugin_handle;
++    if ( (plugin_handle = dlopen(plugin_path, RTLD_LAZY)) == NULL) {
++#ifdef DEBUG
++        itrace("Plugin: can't load plugin %s: %s\n", plugin_path, dlerror());
++#endif
++        return;
++    }
++
++    /* Check if plugin support shell execve method */
++    on_shell_execve_t* plugin_on_shell_execve_handle = dlsym(plugin_handle, on_shell_execve_function_name);
++    if (dlerror() != NULL) {
++        dlclose(plugin_handle);
++
++#ifdef DEBUG
++        itrace("Plugin: can't find on_shell_execve function %s: %s\n", plugin_path, dlerror());
++#endif
++        return;
++    }
++    
++
++    /* Check if plugin support un-initialization method */
++    plugin_uninit_t* plugin_uninit_handle = dlsym(plugin_handle, plugin_uninit_function_name);
++    if (dlerror() != NULL) {
++        dlclose(plugin_handle);
++
++#ifdef DEBUG
++        itrace("Plugin: can't find plugin_uninit function %s: %s\n", plugin_path, dlerror());
++#endif
++        return;
++    }
++
++    /* Check if plugin support initialization method */
++    plugin_init_t* plugin_init_handle = dlsym(plugin_handle, plugin_init_function_name);
++    if (dlerror() != NULL) {
++        dlclose(plugin_handle);
++
++#ifdef DEBUG
++        itrace("Plugin: can't find plugin_init function %s: %s\n", plugin_path, dlerror());
++#endif
++        return;
++    }
++    else {
++        /* Initialize plugin */
++        plugin_init_handle();
++    }
++
++    /* Add plugin to plugin list */
++    int plugin_load_result = append_plugin(plugin_handle,
++                    plugin_on_shell_execve_handle,
++                    plugin_init_handle,
++                    plugin_uninit_handle);
++
++    if (plugin_load_result == PLUGIN_LOAD_SUCCESS) {
++#ifdef DEBUG
++        itrace("Plugin: plugin %s loaded\n", plugin_path);
++#endif
++    }
++    else {
++        /* Output plugin load error message. */
++        fprintf(stderr, "Plugin: plugin %s load failed, result: %d\n", plugin_path, plugin_load_result);
```

Comment:
>fprintf [](http://example.com/codeflow?start=9&length=7)

We need to add syslog. Since the error is silently ignored, we need syslog to monitor.

Reply:
Fixed.